### PR TITLE
Add the ability to rerun success state task_instance in the Backfill API(#16452)

### PR DESCRIPTION
**Pull request by yyhecust**
---
related: #16452
Rerun failed state task_instance was supported in the original Backfill API, this commit adds a feature to the Backfill API that also supports reruning succeeded state task_instance.
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/cli/cli_parser.py [hunk 1]
```diff
@@ -884,7 +893,9 @@ class GroupCommand(NamedTuple):
             "Run subsections of a DAG for a specified date range. If reset_dag_run option is used, "
             "backfill will first prompt users whether airflow should clear all the previous dag_run and "
             "task_instances within the backfill date range. If rerun_failed_tasks is used, backfill "
-            "will auto re-run the previous failed task instances  within the backfill date range"
+            "will auto re-run the previous failed task instances  within the backfill date range. If "
+            "rerun_succeeded_tasks is used, backfill will auto re-run the previous failed task instances "
```
Comment (ephraimbuddy) [inline airflow/cli/cli_parser.py]:
```suggestion
            "rerun_succeeded_tasks is used, backfill will auto re-run the previous succeeded task instances "
```
I think that's what you want to do, right?

File: airflow/jobs/backfill_job.py [hunk 2]
```diff
@@ -164,6 +165,9 @@ def __init__(  # pylint: disable=too-many-arguments
         :param rerun_failed_tasks: flag to whether to
                                    auto rerun the failed task in backfill
         :type rerun_failed_tasks: bool
+        :param rerun_succeeded_tasks: flag to whether to
+                                   auto rerun the failed task in backfill
```
Comment (ephraimbuddy) [inline airflow/jobs/backfill_job.py]:
```suggestion
                                   auto rerun the succeeded task in backfill
```

File: airflow/jobs/backfill_job.py [hunk 3]
```diff
@@ -471,6 +476,16 @@ def _per_task_process(key, ti, session=None):  # pylint: disable=too-many-return
...
+                elif self.rerun_succeeded_tasks:
+                    # Rerun succeeded tasks
+                    if ti.state in (State.SUCCESS):
+                        self.log.error("Task instance {ti} "
+                                       "with state {state}".format(ti=ti,
+                                                                   state=ti.state))
+                        if key in ti_status.running:
+                            ti_status.running.pop(key)
+                        # Reset the succeeded task in backfill to scheduled state
+                        ti.set_state(State.SCHEDULED, session=session)
```
Comment (ephraimbuddy) [inline airflow/jobs/backfill_job.py]:
```suggestion
                elif self.rerun_succeeded_tasks and ti.state == State.SUCCESS:
                    # Rerun succeeded tasks
                    self.log.error("Task instance {ti} with state {state}".format(ti=ti,
                                                                   state=ti.state))
                    if key in ti_status.running:
                        ti_status.running.pop(key)
                    # Reset the succeeded task in backfill to scheduled state
                    ti.set_state(State.SCHEDULED, session=session)
```

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]

File: airflow/jobs/backfill_job.py [hunk 4]
```diff
@@ -441,7 +446,7 @@ def _per_task_process(key, ti, session=None):  # pylint: disable=too-many-return
 
                 # The task was already marked successful or skipped by a
                 # different Job. Don't rerun it.
-                if ti.state == State.SUCCESS:
+                if ti.state == State.SUCCESS and self.rerun_succeeded_tasks == False:
```
Comment (ephraimbuddy) [inline airflow/jobs/backfill_job.py]:
```suggestion
                if ti.state == State.SUCCESS and not self.rerun_succeeded_tasks:
```

**Comment by ephraimbuddy** [comment]
@ashb please take a look.

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]
Add tests

File: airflow/cli/cli_parser.py (hunk 1 above)
Comment (yyhecust) [reply airflow/cli/cli_parser.py]: yes

**Comment by yyhecust** [comment]
> Add tests

ok

File: airflow/models/dag.py [hunk 5]
```diff
@@ -1662,6 +1662,7 @@ def run(
         verbose=False,
         conf=None,
         rerun_failed_tasks=False,
+        rerun_succeeded_tasks=False,
```
Comment (ephraimbuddy) [inline airflow/models/dag.py:2186]: You should add a test for the change here

Comment (yyhecust) [reply airflow/models/dag.py:2186]: I have add a test_cli_backfill_rerun_succeeded_tasks func in test_dag_command.py，does it meet your requirements?

File: airflow/cli/cli_parser.py [hunk 6]
```diff
@@ -277,6 +277,15 @@ def _check(value):
     ),
     action="store_true",
 )
+ARG_RERUN_SUCCEED_TASKS = Arg(
+    ("--rerun-succeeded-tasks",),
+    help=(
+        "if set, the backfill will auto-rerun "
+        "all the succeeded tasks for the backfill date range "
+        "instead of throwing exceptions"
```
Comment (ephraimbuddy) [inline airflow/cli/cli_parser.py]: Is this likely to throw exceptions if not set?

File: airflow/models/dag.py (hunk 5 above)
Comment (ephraimbuddy) [reply airflow/models/dag.py:2186]: Good. You also need to add this kind of test :
https://github.com/apache/airflow/blob/1e1b212417e089b43474bc19eec05b654afe98fb/tests/jobs/test_backfill_job.py#L574-L576
for the rerun success task

File: airflow/cli/cli_parser.py (hunk 6 above)
Comment (yyhecust) [reply airflow/cli/cli_parser.py]: if not set, it won't throw exceptions, the program will set rerun_succeeded_tasks to false by default.

File: airflow/models/dag.py (hunk 5 above)
Comment (yyhecust) [reply airflow/models/dag.py:2186]: OK, I have added the test in recent commit (a85e664)

File: airflow/cli/cli_parser.py (hunk 6 above)
Comment (ephraimbuddy) [reply airflow/cli/cli_parser.py]: So the part you said `instead of throwing exceptions should be removed.

**Comment by uranusjr** [comment]
It seems the rebase went wrong, please try again.

File: tests/jobs/test_backfill_job.py [hunk 7]
```diff
@@ -56,47 +54,48 @@
...
+        return dag
 
-    def test_unfinished_dag_runs_set_to_failed(self):
-        dag = self._get_dummy_dag('dummy_dag')
+    def _times_called_with(self, method, class_):
+        count = 0
+        for args in method.call_args_list:
+            if isinstance(args[0][0], class_):
+                count += 1
+        return count
```
Comment (uranusjr) [inline tests/jobs/test_backfill_job.py]:
```suggestion
        return sum(1 for args in method.call_args_list if isinstance(args[0][0], class_))
```

A common trick :)

File: airflow/cli/cli_parser.py (hunk 6 above)
Comment (yyhecust) [reply airflow/cli/cli_parser.py]: This is written in imitation of the rerun_failed_tasks function

Comment (yyhecust) [reply airflow/cli/cli_parser.py]: and i can remove it, thank you

File: tests/jobs/test_backfill_job.py (hunk 7 above)
Comment (yyhecust) [reply tests/jobs/test_backfill_job.py]: this is a good trick, i will accept it

File: airflow/jobs/backfill_job.py [hunk 8]
```diff
@@ -471,6 +476,14 @@ def _per_task_process(key, ti, session=None):
                             ti_status.running.pop(key)
                         # Reset the failed task in backfill to scheduled state
                         ti.set_state(State.SCHEDULED, session=session)
+                elif self.rerun_succeeded_tasks and ti.state == State.SUCCESS:
+                    # Rerun succeeded tasks
+                    self.log.error("Task instance {ti} with state {state}".format(ti=ti,
+                                                                   state=ti.state))
```
Comment (ephraimbuddy) [inline airflow/jobs/backfill_job.py]:
```suggestion
                    self.log.info("Task instance %s with state %s succeeded", ti,ti.state)
```
This should not be an error log, right?

File: tests/jobs/test_backfill_job.py [hunk 9]
```diff
@@ -37,7 +37,7 @@
     TaskConcurrencyLimitReached,
 )
 from airflow.jobs.backfill_job import BackfillJob
-from airflow.models import DagBag, Pool, TaskInstance as TI
+from airflow.models import DAG, DagBag, Pool, TaskInstance as TI
```
Comment (ephraimbuddy) [inline tests/jobs/test_backfill_job.py]:
```suggestion
from airflow.models import DagBag, Pool, TaskInstance as TI
```

File: tests/jobs/test_backfill_job.py [hunk 10]
```diff
@@ -287,10 +284,7 @@ def test_backfill_conf(self, dag_maker):
         )
         job.run()
 
-        # We ignore the first dag_run created by fixture
-        dr = DagRun.find(
-            dag_id='test_backfill_conf', execution_start_date=DEFAULT_DATE + datetime.timedelta(days=1)
-        )
+        dr = DagRun.find(dag_id='test_backfill_conf')
```
Comment (ephraimbuddy) [inline tests/jobs/test_backfill_job.py]:
```suggestion
        # We ignore the first dag_run created by fixture
        dr = DagRun.find(
            dag_id='test_backfill_conf', execution_start_date=DEFAULT_DATE + datetime.timedelta(days=1)
        )
```
Does this affect your change?

File: tests/jobs/test_backfill_job.py [hunk 11]
```diff
@@ -477,6 +471,8 @@ def test_backfill_pool_not_found(self, dag_maker):
         except AirflowException:
             return
 
+        self.fail()
```
Comment (ephraimbuddy) [inline tests/jobs/test_backfill_job.py]:
```suggestion
```
Any reason for adding this?

File: tests/jobs/test_backfill_job.py [hunk 12]
```diff
@@ -612,6 +640,7 @@ def test_backfill_rerun_upstream_failed_tasks(self, dag_maker):
             op1.set_upstream(op2)
         dag_maker.create_dagrun()
 
+        dag.clear()
```
Comment (ephraimbuddy) [inline tests/jobs/test_backfill_job.py]:
```suggestion
```
??

File: tests/jobs/test_backfill_job.py [hunk 13]
```diff
@@ -719,7 +748,7 @@ def test_backfill_retry_always_failed_task(self, dag_maker):
             start_date=DEFAULT_DATE,
             end_date=DEFAULT_DATE,
         )
-        with pytest.raises(BackfillUnfinished):
+        with self.assertRaises(BackfillUnfinished):
```
Comment (ephraimbuddy) [inline tests/jobs/test_backfill_job.py]:
```suggestion
        with pytest.raises((BackfillUnfinished):
```
We no longer use unittest model on these tests, so this won't work

File: tests/jobs/test_backfill_job.py [hunk 14]
```diff
@@ -740,6 +769,8 @@ def test_backfill_ordered_concurrent_execute(self, dag_maker):
             op3.set_downstream(op4)
         dag_maker.create_dagrun()
 
+        dag.clear()
+
```
Comment (ephraimbuddy) [inline tests/jobs/test_backfill_job.py]:
```suggestion
```

File: tests/jobs/test_backfill_job.py [hunk 15]
```diff
@@ -929,8 +960,11 @@ def run_backfill(cond):
                     dag_maker.create_dagrun(
                         # Existing dagrun that is not within the backfill range
                         run_id=run_id,
+                        state=State.RUNNING,
                         execution_date=DEFAULT_DATE + datetime.timedelta(hours=1),
+                        start_date=DEFAULT_DATE,
```
Comment (ephraimbuddy) [inline tests/jobs/test_backfill_job.py]: dag_maker has a default state=`State.RUNNING` and start_date=DEFAULT_DATE. Why are we adding this? Less code is better...
I don't think you should change this test

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]
I think you had issues with rebasing, some tests were modified and they should not be modified. I have added some suggestions but please remove other changes you made on the tests that are not connected to this PR

File: tests/jobs/test_backfill_job.py (hunk 15 above)
Comment (yyhecust) [reply tests/jobs/test_backfill_job.py]: Yes,I will resolve these problems, thank you

File: airflow/jobs/backfill_job.py [hunk 16]
```diff
@@ -471,6 +476,13 @@ def _per_task_process(key, ti, session=None):
                             ti_status.running.pop(key)
                         # Reset the failed task in backfill to scheduled state
                         ti.set_state(State.SCHEDULED, session=session)
+                elif self.rerun_succeeded_tasks and ti.state == State.SUCCESS:
+                    # Rerun succeeded tasks
+                    self.log.info("Task instance {ti} with state {state}".format(ti=ti, state=ti.state))
```
Comment (ephraimbuddy) [inline airflow/jobs/backfill_job.py]:
```suggestion
                    self.log.info("Task instance %s with state %s", ti, ti.state)
```
For some reasons, we don't use `format` in log messages. We prefer using `%s`. Also, should we make the log to convey a good message, like, Task instance with state ... succeeded. rerunning task? what do you think?

**Review by ephraimbuddy — CHANGES_REQUESTED** [review]

File: airflow/jobs/backfill_job.py (hunk 16 above)
Comment (yyhecust) [reply airflow/jobs/backfill_job.py]: Yes, it will make the feature clearer to users.

Comment (yyhecust) [reply airflow/jobs/backfill_job.py]: I have updated log message in recent commit

File: airflow/jobs/backfill_job.py [hunk 17]
```diff
@@ -478,7 +478,7 @@ def _per_task_process(key, ti, session=None):
                         ti.set_state(State.SCHEDULED, session=session)
                 elif self.rerun_succeeded_tasks and ti.state == State.SUCCESS:
                     # Rerun succeeded tasks
-                    self.log.info("Task instance %s with state %s", ti, ti.state)
+                    self.log.info("Task instance %s with state %s, rerunning task", ti, ti.state)
```
Comment (ephraimbuddy) [inline airflow/jobs/backfill_job.py]:
```suggestion
                    self.log.info("Task instance %s with state %s, rerunning task" succeeded, ti, ti.state)
```
Would that be better?

**Comment by ephraimbuddy** [comment]
@ash, @kaxil what do you think about this PR?

**Comment by yyhecust** [comment]
@ash, @kaxil I am looking forward to your reply

File: airflow/cli/cli_parser.py [hunk 18]
```diff
@@ -909,7 +917,9 @@ class GroupCommand(NamedTuple):
             "Run subsections of a DAG for a specified date range. If reset_dag_run option is used, "
             "backfill will first prompt users whether airflow should clear all the previous dag_run and "
             "task_instances within the backfill date range. If rerun_failed_tasks is used, backfill "
-            "will auto re-run the previous failed task instances  within the backfill date range"
+            "will auto re-run the previous failed task instances  within the backfill date range. If "
```
Comment (uranusjr) [inline airflow/cli/cli_parser.py]:
```suggestion
            "will auto re-run the previous failed task instances within the backfill date range. If "
```

File: airflow/cli/commands/dag_command.py [hunk 19]
```diff
@@ -122,13 +123,15 @@ def dag_backfill(args, dag=None):
...
                 rerun_failed_tasks=args.rerun_failed_tasks,
+                rerun_succeeded_task=args.rerun_succeeded_tasks,
                 run_backwards=args.run_backwards,
             )
         except ValueError as vr:
             print(str(vr))
             sys.exit(1)
 
 
+
```
Comment (uranusjr) [inline airflow/cli/commands/dag_command.py]: I think linter would complain about this extra empty line?

**Comment by uranusjr** [comment]
A couple of nitpicks, lgtm in general.

**Pull request closed without merging**